File: Leetcode/Populating Next Right Pointers in Each Node II.cc
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root) {
            return;
        }
        if (!root -> left && !root -> right) {
            return;
        }
        // --
        if (root -> left && root -> right) {
            root -> left -> next = root -> right;
        }

        TreeLinkNode *parent_next = root -> next;
        TreeLinkNode *child_next = NULL;
        while (parent_next && !child_next) {
            if (parent_next -> left) {
                child_next = parent_next -> left;
            } else if (parent_next-> right) {
                child_next = parent_next -> right;
            }
            
            parent_next = parent_next -> next;
        }
        if (root -> right) {
            root -> right -> next = child_next;
        } else if (root -> left) {
            root -> left -> next = child_next;
        }
        if (root -> right) {
            connect(root -> right);
        }
        if (root -> left) {
            connect(root -> left);
        }
    }
};

File: Leetcode/java/Populating Next Right Pointers in Each Node II.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        while (root!= null) {
            do_connect(root);
            while (root != null && root.left == null && root.right == null) {
                root = root.next;
            }
            if (root == null) {
                break;
            }
            if (root.left != null) {
                root = root.left;
            } else {
                root = root.right;
            }
        }
    }
    public TreeLinkNode do_connect(TreeLinkNode root) {
        if (root == null) {
            return null;
        }
        if (root.left == null && root.right == null) {
            return do_connect(root.next);
        }
        if (root.left != null && root.right != null) {
            root.left.next = root.right;
            root.right.next = do_connect(root.next);
            return root.left;
        }
        if (root.left != null) {
            root.left.next = do_connect(root.next);
            return root.left;
        }
        if (root.right != null) {
            root.right.next = do_connect(root.next);
            return root.right;
        }
        return null;
    }
}
